patches/server/1055-Adding-PredicateChoice-to-Paper-API.patch

```diff
@@ -0,0 +1,210 @@
++                isPotentialIngredient = true;
++                final IntList idList = extraStackingIds.computeIfAbsent(ingredient, ingredient1 -> new IntArrayList());
++                idList.add(id);
++                if (!exactChoice.hasTag())
++                    idList.add(StackedContents.getStackingIndex(exactChoice));
++                idList.sort(IntComparators.NATURAL_COMPARATOR);
++            }
++        }
++
++        if(!isPotentialIngredient)
++            return -1;
+         this.exactChoiceIds.put(exactChoice, id);
+         this.idToExactChoice.put(id, exactChoice);
+         return id;
+@@ -67,6 +80,7 @@ public final class StackedContentsExtraMap {
+     }
+ 
+     public boolean accountStack(final ItemStack stack, final int count) {
++        registerExact(stack);
+         if (!this.exactChoiceIds.isEmpty()) {
+             final int id = this.exactChoiceIds.getInt(stack);
+             if (id >= 0) {
+@@ -74,6 +88,7 @@ public final class StackedContentsExtraMap {
+                 return true;
+             }
+         }
+-        return false;
++        // We always want to use extra map
++        return true;
+     }
+ }
+diff --git a/src/main/java/net/minecraft/world/entity/player/StackedContents.java b/src/main/java/net/minecraft/world/entity/player/StackedContents.java
+index 26b236a764177ac16d53f5cbaf83d3e21d015ebc..82eda864013cbebf94a28e9de49efc23c562238d 100644
+--- a/src/main/java/net/minecraft/world/entity/player/StackedContents.java
++++ b/src/main/java/net/minecraft/world/entity/player/StackedContents.java
+@@ -24,7 +24,8 @@ public class StackedContents {
+     @Nullable public io.papermc.paper.inventory.recipe.StackedContentsExtraMap extrasMap = null; // Paper
+ 
+     public void accountSimpleStack(ItemStack stack) {
+-        if (this.extrasMap != null && stack.hasTag() && this.extrasMap.accountStack(stack, Math.min(64, stack.getCount()))) return; // Paper - max of 64 due to accountStack method below
++        // We always want to use extra map!
++        if (this.extrasMap != null /*&& stack.hasTag()*/ && this.extrasMap.accountStack(stack, Math.min(64, stack.getCount()))) return; // Paper - max of 64 due to accountStack method below
+         if (!stack.isDamaged() && !stack.isEnchanted() && !stack.hasCustomHoverName()) {
+             this.accountStack(stack);
+         }
+@@ -39,7 +40,8 @@ public class StackedContents {
+         if (!stack.isEmpty()) {
+             int i = getStackingIndex(stack);
+             int j = Math.min(maxCount, stack.getCount());
+-            if (this.extrasMap != null && stack.hasTag() && this.extrasMap.accountStack(stack, j)) return; // Paper - if an exact ingredient, don't include it
++            // We always want to use extra map!
++            if (this.extrasMap != null /*&& stack.hasTag()*/ && this.extrasMap.accountStack(stack, j)) return; // Paper - if an exact ingredient, don't include it
+             this.put(i, j);
```

Comment:
The extras map is now just completely replacing the logic for accounting stacks. Probably good to mention that, rename the field accordingly.